api/src/test/kotlin/no/nav/helsearbeidsgiver/inntektsmelding/api/lagreselvbestemtim/LagreSelvbestemtImRouteKtTest.kt

```diff
@@ -32,29 +35,192 @@ class LagreSelvbestemtImRouteKtTest : ApiTest() {
 
         coEvery { mockRedisPoller.hent(any()) } returnsMany listOf(
             harTilgangResultat,
-            ResultJson(
-                success = selvbestemtId.toJson()
-            ).toJson(ResultJson.serializer())
+            Mock.successResult(selvbestemtId)
         )
 
         val response = post(path, mockSkjemaInntektsmeldingSelvbestemt(), SkjemaInntektsmeldingSelvbestemt.serializer())
 
-        assertEquals(HttpStatusCode.OK, response.status)
-        assertEquals(LagreSelvbestemtImResponse(selvbestemtId = selvbestemtId), response.bodyAsText().fromJson(LagreSelvbestemtImResponse.serializer()))
+        val actualJson = response.bodyAsText()
+
+        response.status shouldBe HttpStatusCode.OK
+        actualJson shouldBe Mock.successResponseJson(selvbestemtId)
+    }
+
+    @Test
+    fun `feil i request body gir 400-feil`() = testApi {
+        val expectedFeilmelding = "Feil under serialisering."
+
+        coEvery { mockRedisPoller.hent(any()) } returns harTilgangResultat
+
+        val response = post(path, "ikke et skjema", String.serializer())
+
+        val actualJson = response.bodyAsText()
+
+        response.status shouldBe HttpStatusCode.BadRequest
+        actualJson shouldBe Mock.failureResponseJson(expectedFeilmelding)
+    }
+
+    @Test
+    fun `valideringsfeil gir 400-feil`() = testApi {
+        val expectedFailureResponseJson = """
+        {
+            "valideringsfeil": [
+                "Sykmeldingsperioder må fylles ut",
+                "Beløp må være større eller lik 0",
+                "Refusjonsbeløp må være mindre eller lik inntekt"
+            ],
+            "error": "Feil under validering."
+        }
+        """.removeJsonWhitespace()
+
+        coEvery { mockRedisPoller.hent(any()) } returns harTilgangResultat
+
+        val skjemaMedFeil = mockSkjemaInntektsmeldingSelvbestemt().let {
+            it.copy(
+                sykmeldingsperioder = emptyList(),
+                inntekt = it.inntekt.copy(
+                    beloep = -1.0
+                )
+            )
+        }
+
+        val response = post(path, skjemaMedFeil, SkjemaInntektsmeldingSelvbestemt.serializer())
+
+        val actualJson = response.bodyAsText()
+
+        response.status shouldBe HttpStatusCode.BadRequest
+        actualJson shouldBe expectedFailureResponseJson
+    }
+
+    @Test
+    fun `manglende tilgang gir 500-feil`() = testApi {
+        coEvery { mockRedisPoller.hent(any()) } returns ikkeTilgangResultat
+
+        val response = post(path, mockSkjemaInntektsmeldingSelvbestemt(), SkjemaInntektsmeldingSelvbestemt.serializer())
+
+        val actualJson = response.bodyAsText()
+
+        response.status shouldBe HttpStatusCode.InternalServerError
+        actualJson shouldBe "\"Error 500: no.nav.helsearbeidsgiver.inntektsmelding.api.auth.ManglerAltinnRettigheterException\""
```

Comment:
Burde ikke denne egentlig vært forbidden?

Reply:
Ja, det gir mening, men da burde vi kanskje ha det slik for alle kall som feiler tilgangssjekken?

Reply:
Ja det burde vel være likt.
Trodde vi hadde 403 på forespurt inntektsmelding.
Men det er vel forskjellig i selvbestemt tydeligvis.